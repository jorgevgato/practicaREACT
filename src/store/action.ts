import type { Advert } from "../pages/adverts/types";

type AuthLogin = {
  type: "auth/login";
};

type AuthLogout = {
  type: "auth/logout";
};

type AdvertsLoaded = {
  type: "adverts/loaded";
  payload: Advert[];
};

type AdvertsCreated = {
  type: "adverts/created";
  payload: Advert;
};

export const authLogin = (): AuthLogin => ({
  type: "auth/login",
});

export const authLogout = (): AuthLogout => ({
  type: "auth/logout",
});

export const advertsLoaded = (adverts: Advert[]): AdvertsLoaded => ({
  type: "adverts/loaded",
  payload: adverts,
});

export const advertsCreated = (advert: Advert): AdvertsCreated => ({
  type: "adverts/created",
  payload: advert,
});

export type Actions = AuthLogin | AuthLogout | AdvertsLoaded | AdvertsCreated;
