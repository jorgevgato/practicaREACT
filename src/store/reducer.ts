import type { Advert } from "../pages/adverts/types";
import type { Actions } from "./action";

type State = {
  auth: boolean;
  adverts: Advert[];
};

const defaultState: State = {
  auth: false,
  adverts: [],
};

export function reducer(state = defaultState, action: Actions): State {
  switch (action.type) {
    case "auth/login":
      return { ...state, auth: true };
    case "auth/logout":
      return { ...state, auth: false };
    case "adverts/loaded":
      return { ...state, adverts: action.payload };
    case "adverts/created":
      return { ...state, adverts: [...state.adverts, action.payload] };
    default:
      return state;
  }
}
