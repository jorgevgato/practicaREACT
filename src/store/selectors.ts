import type { RootState } from ".";
import type { Advert } from "../pages/adverts/types";

export const getIsLogged = (state: RootState) => state.auth;

export const getAdverts = (state: RootState) => state.adverts;

export const getTags = (state: RootState): string[] => state.tags;

export const getDetail = (state: RootState): Advert | null => state.detail;
