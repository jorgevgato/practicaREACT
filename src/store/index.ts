import { combineReducers, createStore } from "redux";
import * as reducers from "./reducer";
import { composeWithDevTools } from "@redux-devtools/extension";
import { useDispatch, useSelector } from "react-redux";

const rootReducer = combineReducers(reducers);

export default function configureStore(
  preloadedState: Partial<reducers.State>,
) {
  const store = createStore(
    rootReducer,
    preloadedState as never,
    /* // @ts-expect-error: import devtools extension
    window.REDUX_DEVTOOLS_EXTENSION &&
      // @ts-expect-error: import devtools extension
      window.REDUX_DEVTOOLS_EXTENSION(), */
    composeWithDevTools(),
  );
  return store;
}

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof configureStore>;
export type AppDispatch = AppStore["dispatch"];
/* export type AppGetState = AppStore["getState"]; */

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
