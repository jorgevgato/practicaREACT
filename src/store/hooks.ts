import { useAppDispatch, useAppSelector } from ".";
import { authLogin, authLogout } from "./action";
import { getIsLogged } from "./selectors";

export function useAuth() {
  return useAppSelector(getIsLogged);
}

export function useLoginAction() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(authLogin());
  };
}

export function useLogoutAction() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(authLogout());
  };
}
