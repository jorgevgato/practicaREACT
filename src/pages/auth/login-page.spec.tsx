import { describe, test, expect, vi, beforeEach } from "vitest";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import LoginPage from "./login-page";

vi.mock("../../store/hooks", () => ({
  useLoginAction: vi.fn(),
}));

import { useLoginAction } from "../../store/hooks";

vi.mock("react-router", async () => {
  const actual =
    await vi.importActual<typeof import("react-router")>("react-router");
  return {
    ...actual,
    useLocation: () => ({ state: {} }),
    useNavigate: () => vi.fn(),
  };
});

describe("LoginPage component", () => {
  const mockStore = configureMockStore();
  const store = mockStore({
    auth: false,
    adverts: [],
    tags: [],
    detail: null,
  });

  const mockedLoginAction = useLoginAction as unknown as ReturnType<
    typeof vi.fn
  >;

  beforeEach(() => {
    mockedLoginAction.mockReset();
  });

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>,
    );

  test("should render LoginPage and match snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
