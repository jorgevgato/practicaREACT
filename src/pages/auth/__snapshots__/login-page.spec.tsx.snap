// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage component > should render LoginPage and match snapshot 1`] = `
<div>
  <div
    class="main-container"
  >
    <div
      class="login-page"
    >
      <h1>
        Inicia sesión
      </h1>
      <form>
        <label>
          Email
          <input
            name="email"
            type="text"
            value=""
          />
        </label>
        <label>
          Password
          <input
            name="password"
            type="password"
            value=""
          />
        </label>
        <button
          disabled=""
          type="submit"
        >
          Iniciar sesión
        </button>
        <label
          class="remember-check"
        >
          <input
            type="checkbox"
          />
          <span>
            Recordar contraseña
          </span>
        </label>
        <a
          data-discover="true"
          href="/"
        >
           
          <img
            alt="Ir al inicio"
            class="login-logo"
            src="/src/assets/react.svg"
          />
        </a>
      </form>
    </div>
  </div>
</div>
`;
